### PR TITLE
Adds permissions parser for string/octal

### PR DESCRIPTION
**Pull request by reinhapa**
Signed-off-by: Patrick Reinhart <patrick@reini.net>

### CODE REVIEW

File: src/main/java/org/xmldb/api/security/Permissions.java [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+          break;
+        case UNSET_CHAR:
+          break;
+        default:
+          throw new IllegalArgumentException("Unrecognised mode char '" + c + "'");
+      }
+    }
+  }
+
+  private static void setPermission(EnumSet<Permission> permissions, int index, Permission owner,
```
Comment (adamretter) [inline src/main/java/org/xmldb/api/security/Permissions.java:342]: I don't understand this function - it appears to take 3 permission args, but will only ever use 1 of them. Why not just have 1 permission arg?

File: src/main/java/org/xmldb/api/security/Permissions.java [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+  }
+
+
+  /**
+   * Parses the permissions from the given octal mode value.
+   * 
+   * @param mode the octal permission mode
+   * @return the set of parsed permissions
+   */
+  public static Set<Permission> ofOctal(int mode) {
```
Comment (adamretter) [inline src/main/java/org/xmldb/api/security/Permissions.java:137]: Perhaps `fromOctal` is a better name?

File: src/main/java/org/xmldb/api/security/Permissions.java [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+   * <ol>
+   * <li>Unix Symbolic format as given to 'chmod' on Unix/Linux</li>
+   * <li>Simple Symbolic format e.g. "rwxr-xr-x"</li>
+   * </ol>
+   * 
+   * @param modeStr the mode to be parsed
+   * @return the set of parsed permissions
+   * @throws IllegalArgumentException if the given mode string is in the wrong format
+   */
+  public static Set<Permission> ofModeString(String modeStr) {
```
Comment (adamretter) [inline src/main/java/org/xmldb/api/security/Permissions.java:114]: Perhaps `fromModeString` is a better name?

File: src/main/java/org/xmldb/api/security/Permissions.java [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+import java.util.regex.Pattern;
+
+/**
+ * Parser for {@link Permission} values.
+ * 
+ * This class provides static methods to parse string and integer based permission modes into a set
+ * of permissions.
+ */
+public final class Permissions {
+  private static final char READ = 04;
```
Comment (adamretter) [inline src/main/java/org/xmldb/api/security/Permissions.java:69]: I am a little unsure why these are `char`?

File: src/main/java/org/xmldb/api/security/Permissions.java (hunk 1 above)
Comment (reinhapa) [reply src/main/java/org/xmldb/api/security/Permissions.java:342]: The `permissions` argument contains the set of all parsed permissions, the `index` defines the which of the three passed in permissions (`owner`, `group` or 'others`) representing the correct _read_, _write_ or _execute_ permission.

File: src/main/java/org/xmldb/api/security/Permissions.java (hunk 4 above)
Comment (reinhapa) [reply src/main/java/org/xmldb/api/security/Permissions.java:69]: well just to use less memory... could be even of type `byte`

File: src/main/java/org/xmldb/api/security/Permissions.java (hunk 2 above)
Comment (reinhapa) [reply src/main/java/org/xmldb/api/security/Permissions.java:137]: I have no problem to go with `fromModeOctal` if this is more suited.

File: src/main/java/org/xmldb/api/security/Permissions.java (hunk 3 above)
Comment (reinhapa) [reply src/main/java/org/xmldb/api/security/Permissions.java:114]: I have no problem to go with `fromModeString` if this is more suited.

File: src/main/java/org/xmldb/api/security/Permissions.java (hunk 4 above)
Comment (adamretter) [reply src/main/java/org/xmldb/api/security/Permissions.java]: I doubt in practice there will be any difference in memory use here ;-)

File: src/main/java/org/xmldb/api/security/Permissions.java (hunk 1 above)
Comment (adamretter) [reply src/main/java/org/xmldb/api/security/Permissions.java:342]: I am, not sure I understand... The following code would do the same wouldn't it?

```java
private static void setPermission(EnumSet<Permission> permissions, int index, Permission permission) {
      Permission group, Permission others) {
    switch (index) {
      case 0:
        permissions.add(permission);
        break;
      case 3:
        permissions.add(permission);
        break;
      case 6:
        permissions.add(permission);
        break;
      default:
    }
  }
}
```

**Pull request merged**